packages/labs/ssr-react/src/node/jsx-runtime.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2022 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/**
+ * @fileoverview To serve as JSX import source for runtime JSX transforms in
+ * production mode. For use in servers.
+ */
+
+// eslint-disable-next-line import/extensions
+import * as ReactJSXRuntime from 'react/jsx-runtime';
+import {createElement, type ReactNode, type ElementType} from 'react';
+import {isCustomElement} from '../lib/utils.js';
+import {renderShadowContents} from '../lib/node/render-shadow-contents.js';
+
+export const Fragment = ReactJSXRuntime.Fragment;
+
+export const jsx = <P extends {children?: ReactNode}>(
+  type: ElementType<P>,
+  props: P,
+  key: string | undefined
+) => {
+  if (isCustomElement(type)) {
+    const {shadowContents, elementAttributes, templateAttributes} =
+      renderShadowContents(type, props);
+
+    if (shadowContents) {
+      const templateShadowRoot = createElement('template', {
+        ...templateAttributes,
+        dangerouslySetInnerHTML: {
+          __html: [...shadowContents].join(''),
+        },
+      });
+
+      // Call `jsxs` instead of `jsx` so that React doesn't incorrectly
+      // interpret that the children array was dynamically made and warn about
+      // missing keys
+      return ReactJSXRuntime.jsxs(type, {
+        ...props,
+        ...elementAttributes,
+        children: [templateShadowRoot, props.children],
+      });
+    }
+  }
+
+  return ReactJSXRuntime.jsx(type, props, key);
+};
+
+export const jsxs = <P extends {children: ReactNode[]}>(
+  type: ElementType<P>,
+  props: P,
+  key: string | undefined
+) => {
+  if (isCustomElement(type)) {
+    const {shadowContents, elementAttributes, templateAttributes} =
+      renderShadowContents(type, props);
+
+    if (shadowContents) {
+      const templateShadowRoot = createElement('template', {
+        ...templateAttributes,
+        dangerouslySetInnerHTML: {
+          __html: [...shadowContents].join(''),
+        },
+      });
+
```

Comment:
This code is duplicated 4x, is there a way to factor it better? This can't all be just `wrapCreateElement` I guess because there are slight differences in how children, keys, etc. are handled between the different fns, but maybe this inner part could be factored?

Reply:
The `renderShadowContents()` (or maybe now-to-be-called `renderCustomElement()`) was as much commonality I could pull out at the time. Though I think there might be a better way to deduplicate all of this. I'm thinking of actually creating a custom React element to handle the adding of template shadowroot so it would not need to have bespoke handling of the different transforms. I'll see if this idea pans out and push up a commit with it.

Reply:
@augustjk I would be tempted to TODO the deduplication and try a component in a follow-on PR.

Reply:
Yeah, it would just be implementation details and not affect the outward API so seems okay to punt that. @aomarks had the same thought while discussing offline.